File: backend/Shelfie/Shelfie.Services/Database/ReadingChallenge.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Shelfie.Models.Enums;

namespace Shelfie.Services.Database
{
    public class ReadingChallenge : BaseEntity
    {
        public int UserId { get; set; }

        [ForeignKey("UserId")]
        public User User { get; set; } = null!;

        [Required(ErrorMessage = "Challenge name is required.")]
        [MaxLength(100, ErrorMessage = "Challenge name cannot exceed 100 characters.")]
        public string ChallengeName { get; set; } = string.Empty;

        [MaxLength(1000, ErrorMessage = "Description cannot exceed 1000 characters.")]
        public string Description { get; set; } = string.Empty;

   
        [Column(TypeName = "varchar(20)")]
        [Required]
        public GoalTypeEnum GoalType { get; set; }

        [Range(1, int.MaxValue, ErrorMessage = "GoalAmount must be a positive number.")]
        public int GoalAmount { get; set; }

        [Required(ErrorMessage = "Start date is required.")]
        public DateOnly StartDate { get; set; }

        [Required(ErrorMessage = "End date is required.")]
        [CustomValidation(typeof(Validations.Validations), nameof(Validations.Validations.ValidateEndDateAfterStartDate))]
        public DateOnly EndDate { get; set; }

        [Range(0, int.MaxValue, ErrorMessage = "Progress must be a positive number.")]
        public int Progress { get; set; }

        public bool IsCompleted { get; set; }
    }
}

File: Shelfie/ShelfieUI/shelfie_mobile/lib/models/readingChallenge.dart
import 'package:json_annotation/json_annotation.dart';

part 'readingChallenge.g.dart';

@JsonSerializable()
class ReadingChallenge {
  int id;
  String challengeName;
  String description;
  String goalType;
  int goalAmount;
  DateTime startDate;
  DateTime endDate;
  int progress;
  bool isCompleted;
  int userId;
  String? username;

  ReadingChallenge({
    required this.id,
    required this.challengeName,
    required this.description,
    required this.goalType,
    required this.goalAmount,
    required this.startDate,
    required this.endDate,
    required this.progress,
    required this.isCompleted,
    required this.userId,
    this.username,
  });

  factory ReadingChallenge.fromJson(Map<String, dynamic> json) => _$ReadingChallengeFromJson(json);
  Map<String, dynamic> toJson() => _$ReadingChallengeToJson(this);
}

